### ISSUE TITLE
babel-core/register

### ISSUE THREAD

**Issue by kimjuny** [faceyspacey/babel-plugin-universal-import#7]
I'm trying to use this babel-plugin with babel-core/register.

entry.js
```javascript
require('babel-polyfill');
require('babel-core/register')({
  plugins: ["universal-import"],
});
require('./server/server.js');
```

.babelrc
```json
{
  "presets": [
    "es2015", "react", "stage-0"
  ],
  "plugins": [
    "transform-async-to-generator",
    "universal-import"
  ]
}
```

client side routes
```javascript
import React from 'react';
import { Switch, Route } from 'react-router-dom';
import universal from 'react-universal-component';

const Routes = () => (
  <Switch>
    <Route path="/page" component={universal(import('./Page'))} />
  </Switch>
);

export default Routes;

```

error log
```
$ node ./src/entry.js

/Users/erickim/Develop/front/react/test-project/src/client/index.js:39
          return Promise.all([import( /* webpackChunkName: 'Page' */'./Page'), (0, _importCss3.default)('Page')]).then(function (proms) {
                              ^^^^^^
SyntaxError: Unexpected token import
    at createScript (vm.js:53:10)
    at Object.runInThisContext (vm.js:95:10)
    at Module._compile (module.js:543:28)
    at loader (/Users/erickim/Develop/front/react/test-project/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/erickim/Develop/front/react/test-project/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
error Command failed with exit code 1.
```

No error if I'm running a webpack build, though. So I guess this might be a  problem with `babel-core/register`? 🤔🤔

**Comment by faceyspacey** [comment]
It's primarily for webpack, but try *one* these 2 plugins in addition:

https://github.com/airbnb/babel-plugin-dynamic-import-node

https://github.com/airbnb/babel-plugin-dynamic-import-webpack

Without webpack, babel can't deal with `import` without a separate plugin.

Check these docs: https://github.com/faceyspacey/webpack-flush-chunks/blob/master/docs/babel-config.md

If you can't get it working, feel free to hit me on Reactlandia Chat. The challenge with this approach is that it's the "frameworkless approach." So we get to do a lot of the stuff you can do with Next.js, but with far more flexibility. The challenge is all the initial setup we gotta do for this "privilege." But once we get it, it's smooth sailing.

**Comment by kimjuny** [comment]
@faceyspacey thanks, babel-plugin-dynamic-import-webpack works.

```javascript
require('babel-polyfill');
require('babel-core/register')({
  plugins: ["dynamic-import-webpack"],
});
require('./server/server.js');
```

```
{
  "presets": [
    "es2015", "react", "stage-0"
  ],
  "plugins": [
    "transform-async-to-generator",
    "universal-import"
  ]
}
```

**Comment by kimjuny** [comment]
@faceyspacey Hi, I've tried both babel-plugin-dynamic-import-webpack and babel-plugin-dynamic-import-node, there is no `Unexpected token import` error anymore, but server-side render `renderToString()` will always produce a <b>Loading...</b> state dom `<div data-reactroot="" data-reactid="1" data-react-checksum="622727842">Loading...</div>`

**Comment by faceyspacey** [comment]
are you using https://github.com/faceyspacey/webpack-flush-chunks? ...this is a common setback. have u read all the docs?

**Comment by kimjuny** [comment]
@faceyspacey Yes I am using webpack-flush-chunks as setback. Everything works just fine in a production mode(client and server are all compiled by webpack). But when in development mode(client-side compiled by webpack-dev-server and server-side is not built but using `babel-core/register`), there are strange things happening. 👇

```javascript
import React from 'react';
import { StaticRouter } from 'react-router-dom';
import { renderToString } from 'react-dom/server';
import { flushChunkNames } from 'react-universal-component/server';
import webpackFlushChunks from 'webpack-flush-chunks';
import Routes from '../../../client/routes';

const renderMiddleware = async (ctx, next) => {
  const content = renderToString(
    <StaticRouter
      location={ctx.path}
      context={ctx}
    >
      <Routes />
    </StaticRouter>
  );
  console.log(content);
  // <div data-reactroot="" data-reactid="1" data-react-checksum="622727842">Loading...</div>
  // In production mode server-side render was working correctly. 
  // In development mode(using babel-core/register with dynamic-import-webpack, this 'Loading...' will be rendered.

  const chunkNames = flushChunkNames();
  console.log(chunkNames);
  // [ 'default' ]

  const { js } = webpackFlushChunks(ctx.state.webpackStats.toJson(), { chunkNames });
  // FLUSH CHUNKS]: Unable to find default in Webpack chunks. Please check usage of Babel plugin.
  // THIS IS WEIRD: If no 'default' webpack chunks are found, then why scripts exist?
  console.log(js.toString());
  // <script type='text/javascript' src='/build/client/bootstrap.dbda53d89a5bda9ebd84.js'></script>
  // <script type='text/javascript' src='/build/client/main.175fd60bb1854601ddc3.js'></script>

  ctx.body = `
    <!doctype html>
    <html>
      <head>
        <title>koa-react-universal</title>
      </head>
      <body>
        <div id="root">${content}</div>
        ${js}
      </body>
    </html>
  `;

  await next();
};

export default renderMiddleware;
```

**Comment by faceyspacey** [comment]
I'm gonna need a repro.

Have you checked the docs folder in webpack flush chunks. Just last night I added a link to the old boilerplate repos that supported babel. They may help. you u will see the old strategy that didn't involve a plugin. Doing that might solve things.

If you can provide me a repro I'd be happy to check it for you and make any necesssary fixes. I havent spent as much time with a babel server as I have in the past. You may even be able to make a pr to the old babel boilerplate to replicate your issue. It might be as simple as implementing the new babel plugin and removing the manual way from the past.

**Comment by kimjuny** [comment]
@faceyspacey Hi, this is the repro: https://github.com/kimjuny/koa-react-universal.

prod logs
```
➜  koa-react-universal git:(master) ✗ yarn prod
yarn prod v0.24.4
$ npm run clean && cross-env NODE_ENV=production node ./src/production

> @ clean /Users/erickim/Develop/github/koa-react-universal
> rm -rf build

[SYS] webpack building.
[SYS] webpack build finished.
[SYS] server started at port 3008
<div data-reactroot="" data-reactid="1" data-react-checksum="-1178726173">Home</div>
[ 'containers/Home' ]
<script type='text/javascript' src='/build/client/bootstrap.dbda53d89a5bda9ebd84.js'></script>
<script type='text/javascript' src='/build/client/containers/Home.36528153ca3888acc340.js'></script>
<script type='text/javascript' src='/build/client/main.175fd60bb1854601ddc3.js'></script>
```

dev logs
```
➜  koa-react-universal git:(master) ✗ yarn dev
yarn dev v0.24.4
$ cross-env NODE_ENV=development node ./src/development
[SYS] server started at port 3008
(node:39093) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: require.ensure is not a function
(node:39093) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
[SYS] webpack build finished
webpack built 85bd4c01ed204c5ec3d4 in 9687ms
<div data-reactroot="" data-reactid="1" data-react-checksum="622727842">Loading...</div>
[ 'default' ]
[FLUSH CHUNKS]: Unable to find default in Webpack chunks. Please check usage of Babel plugin.
<script type='text/javascript' src='/build/client/bootstrap.dbda53d89a5bda9ebd84.js'></script>
<script type='text/javascript' src='/build/client/main.175fd60bb1854601ddc3.js'></script>
```

**Comment by faceyspacey** [comment]
Ok, i reviewed it and spent some time twiddling with it. Here's what works. do this at the top of *server/development.js* instead:

```js
require('babel-register')({
  "presets": [
    "es2015",
    "react",
    "stage-0"
  ],
  "plugins": [
    "transform-async-to-generator",
    "universal-import",
    "syntax-dynamic-import"
  ]
});
```

Notice I added the presets too. I think only if you provide no options does it use your babelrc. 

Also note, babel-core/register works just as fine. babel-register is just the newer way to do things.

so you need to add `babel-plugin-syntax-dynamic-import`. the order doesn't matter. This babel plugin just makes sure the compiler doesn't break when it encounters `import()`, whereas the other 2 actually overwrite `import()` to be `require.ensure` or `require()`. And since this will happen before my plugin (again, regardless of the order, for whatever reason), my plugin won't be able to operate.

So syntax-dynamic-import leaves `import()` in tack, but if it's called server-side, it wont be replaced with anything that can work in node. That's fine--as react-universal-component will call one of the other generated options: `resolve` or `path`. And it will do so synchronously. It only calls `load()` which contains `import()` on the client. And only when the module cant be required synchronously.

**Comment by faceyspacey** [comment]
if you run into any problems. delete your node_modules folder and reinstall it. it should be fine then.

**Comment by faceyspacey** [comment]
the doc has also been updated for babel-config. thanks for drawing my attention to it:

https://github.com/faceyspacey/webpack-flush-chunks/blob/master/docs/babel-config.md

**Comment by kimjuny** [comment]
@faceyspacey ok... after days of struggling...

### renderToString is always `Loading...` when using babel server

This is what I was doing(wrong):

```javascript
universal(() => import('./Test'))
```

This is right:
```javascript
universal(() => import(/* webpackChunkName: 'Test' */ './Test'), {
  path: require('path').resolve(__dirname, './Test'),
  resolve: () => require.resolveWeak('./Test'),
  chunkName: 'Test',
})
```

Or if u're using [universal-import](https://github.com/faceyspacey/babel-plugin-universal-import), you can do this:
```javascript
universal(import('./Test'), {
  path: require('path').resolve(__dirname, './Test'),
  resolve: () => require.resolveWeak('./Test'),
  chunkName: 'Test',
})
```

And I'm still working on finding a way to simplify this. 👆

### [FLUSH CHUNKS]: Unable to find xxx in Webpack chunks. Please check usage of Babel plugin.

The key point is: 1. use `babel-plugin-dynamic-import-webpack` for the server-side 2. exclude `babel-plugin-dynamic-import-webpack` from client-side!

So this is what I do.

For the server entry:
```javascript
require('babel-polyfill');
require('babel-core/register')({
  plugins: [
    'dynamic-import-webpack', // use this only in the server entry
    [
      'inline-replace-variables', {
        '__ROOT__': require('path').resolve(__dirname, '../'),
      },
    ],
  ],
});
require('./server/server.js').default();
```

And my `.babelrc`
```json
{
  "presets": [
    "es2015",
    "react",
    "stage-0"
  ],
  "plugins": [
    "universal-import",
    "transform-async-to-generator"
  ]
}
```

I hope this helps if someone encounter same problem.

**Comment by faceyspacey** [comment]
um, did u see what i wrote above and the PR i just did to ur repo.

**Comment by faceyspacey** [comment]
but overall your approach is correct. client side you can't apply the stuff that strips `import()` since `import()` is exclusively used on the client. 

On the server, it doesnt really matter what u do as long as babel compiles, since `import()` will never be used. The synchronous require version is used instead.

**Comment by kimjuny** [comment]
@faceyspacey I had checked out your PR just now! After `yarn` and `yarn dev` I saw an error and couldn't start up:

```
➜  koa-react-universal git:(master) ✗ yarn dev
yarn dev v0.24.4
$ cross-env NODE_ENV=development node ./src/development
/Users/erickim/Develop/github/fork/koa-react-universal/src/client/routes/index.jsx:43
      return Promise.all([import( /* webpackChunkName: 'containers/Home' */'../containers/Home'), (0, _importCss3.default)('containers/Home')]).then(function (proms) {
                          ^^^^^^
SyntaxError: Unexpected token import
    at createScript (vm.js:53:10)
    at Object.runInThisContext (vm.js:95:10)
    at Module._compile (module.js:543:28)
    at loader (/Users/erickim/Develop/github/fork/koa-react-universal/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) [as .jsx] (/Users/erickim/Develop/github/fork/koa-react-universal/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
error Command failed with exit code 1.
```

It could start-up again if I remove "universal-import" from `babel-core/register`, but the problem remains(renderToString is always Loading... && [FLUSH CHUNKS]: Unable to find xxx in Webpack chunks. Please check usage of Babel plugin). I'll let you know if I make any further progress, I'm so sorry I've taken too much of your time!

**Comment by faceyspacey** [comment]
did u kill ur node_modules folder.

**Comment by faceyspacey** [comment]
make sure not to merge it too. try what i PRed fresh. it uses the raw universal(import()) like u originally had. i.e. without the options.

**Comment by kimjuny** [comment]
@faceyspacey Yes I just checked out from your repo https://github.com/faceyspacey/koa-react-universal, there are no `node_modules`.

```
➜  koa-react-universal git:(master) ✗ rm -rf build
➜  koa-react-universal git:(master) ✗ rm -rf node_modules
➜  koa-react-universal git:(master) ✗ yarn
yarn install v0.24.4
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...
✨  Done in 11.00s.
➜  koa-react-universal git:(master) yarn dev
yarn dev v0.24.4
$ cross-env NODE_ENV=development node ./src/development
/Users/erickim/Develop/github/fork/koa-react-universal/src/client/routes/index.jsx:43
      return Promise.all([import( /* webpackChunkName: 'containers/Home' */'../containers/Home'), (0, _importCss3.default)('containers/Home')]).then(function (proms) {
                          ^^^^^^
SyntaxError: Unexpected token import
    at createScript (vm.js:53:10)
    at Object.runInThisContext (vm.js:95:10)
    at Module._compile (module.js:543:28)
    at loader (/Users/erickim/Develop/github/fork/koa-react-universal/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) [as .jsx] (/Users/erickim/Develop/github/fork/koa-react-universal/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
error Command failed with exit code 1.
```

**Comment by faceyspacey** [comment]
u gotta do `yarn` first to install it.

**Comment by faceyspacey** [comment]
thats very weird. that it worked for me and not you.

fyi, the only reason it's working server-side with the babel plugin is cuz ur providing ur own options. the babel plugin is completely destroyed by the -node or -webpack dynamic-import plugins.

**Comment by faceyspacey** [comment]
in the future FYI we may completely stop supporting the babel option. U may really wanna rethink going the webpack server route now too. There's a lot to gain and it seems most people have switched.

**Comment by faceyspacey** [comment]
either way, i know what needs to be done. im glad u got it working. ur solution is a fine solution. its what we did before the plugin. lets consider this completely closed unless u need something else.

**Comment by kimjuny** [comment]
@faceyspacey Yes I'm rethinking this route. I was trying to use babel-node+nodemon or node+chokidar for development because there was no koa-middleware like [webpack-hot-server-middleware](https://github.com/60frames/webpack-hot-server-middleware). Although problems are solved temporarily(still trying to figure out way to simplify universal import with options), I can't guarentee there won't be any other weird problems (so I can keep bugging you 😂), so now I'm considering making one(hot server middleware for koa), it shouldn't be too difficult.

**Comment by faceyspacey** [comment]
def do it! ...im interested into looking more into Koa myself.

Also, ive just started reworking the babel plugin for a babel server. It should be up soon so it wont be causing these problems.

**Comment by faceyspacey** [comment]
I updated the plugin to have a `babelServer: true` option, which should only be used on the server. Here's the docs:

https://github.com/faceyspacey/webpack-flush-chunks/blob/master/docs/babel-config.md

Thanks for drawing my attention to this. Let me know if you ever make that koa middleware--it will be very useful. I need to research if it's worth moving these demos to Koa. Feel free to let me know anything else you like about Koa.

**Comment by kimjuny** [comment]
@faceyspacey I have made some progress. No weird problem so far.

* [koa-webpack-server](https://github.com/kimjuny/koa-webpack-server)
* [koa-react-universal](https://github.com/kimjuny/koa-react-universal)

I'll let you know if these are all ready.

**Comment by faceyspacey** [comment]
Awesome sauce!

**Comment by hrasoa** [comment]
Hi, same as @kimjuny https://github.com/faceyspacey/babel-plugin-universal-import/issues/7#issuecomment-322123215 on dev environment: I got the ```Loading...``` message .

Then when I added ```babel-plugin-dynamic-import-node``` I have now this error on every async components:

```
ERROR in ./src/containers/AsyncProfilePage/index.js
    Module build failed: TypeError: /var/www/src/containers/AsyncProfilePage/index.js: Property value expected type of string but got null
        at Object.validate (/var/www/node_modules/babel-core/node_modules/babel-types/lib/definitions/index.js:161:13)
        at validate (/var/www/node_modules/babel-core/node_modules/babel-types/lib/index.js:505:9)
        at Object.builder (/var/www/node_modules/babel-core/node_modules/babel-types/lib/index.js:466:7)
        at idOption (/var/www/node_modules/babel-plugin-universal-import/index.js:99:51)
        at PluginPass.Import (/var/www/node_modules/babel-plugin-universal-import/index.js:182:13)
        at newFn (/var/www/node_modules/babel-core/node_modules/babel-traverse/lib/visitors.js:276:21)
        at NodePath._call (/var/www/node_modules/babel-core/node_modules/babel-traverse/lib/path/context.js:76:18)
        at NodePath.call (/var/www/node_modules/babel-core/node_modules/babel-traverse/lib/path/context.js:48:17)
        at NodePath.visit (/var/www/node_modules/babel-core/node_modules/babel-traverse/lib/path/context.js:105:12)
        at TraversalContext.visitQueue (/var/www/node_modules/babel-core/node_modules/babel-traverse/lib/context.js:150:16)
     @ ./src/components/App.js 31:14-61
     @ ./server/serverRenderer.js
```

Here is my repo:
https://github.com/hrasoa/react-pwa/tree/universal-comp

**Comment by kimjuny** [comment]
@hrasoa If you are using `babel-core/register` or `babel-server`, you can try this [solution](https://github.com/faceyspacey/babel-plugin-universal-import/issues/7#issuecomment-322123215). Also, you may try `babelServer: true` as @faceyspacey commented [here](https://github.com/faceyspacey/babel-plugin-universal-import/issues/7#issuecomment-322244415).

**Comment by hrasoa** [comment]
@kimjuny  yes I saw your solution, but shouldn't ```babel-plugin-dynamic-import-node``` replace the ```import()``` stuff with synchronous ```require``` on the server side ?

Edit: I'm not using none of babel-core/register and babel-server

**Comment by hrasoa** [comment]
For the js error i rolled back ```babel-plugin-dynamic-import-node``` to fixed ```1.0.2```

@kimjuny as u suggested I changed this and it works now:
```jsx
export default universal(import('./About'));
````
to
```jsx
export default universal(import('./About'), {
  path: require('path').resolve(__dirname, './About'),
  resolve: () => require.resolveWeak('./About'),
  chunkName: 'About'
});
```
